### ISSUE TITLE
[Feature] gen_invalid_dependency should be removed

### ISSUE THREAD

**Issue by arcanis** [yarnpkg/berry#122]
**Describe the context**

The gen_invalid_dependency rules cannot be fixed. In fact they don't even contain enough information to potentially lead to a fix *unless the users encode them into the reason*. Even an interactive mode wouldn't allow to fix the problems.

Additionally, it might be confusing for our users to figure out whether they want to use the `gen_enforced_dependency_range` predicate or the `gen_invalid_dependency` one.

**Describe the solution you'd like**

- The `gen_invalid_dependency` predicate should disappear.

- The engine should be smart enough to detect when `gen_enforced_dependency_range` would have multiple conflicting solutions for a same range.

Doing this would have several advantages:

- No reason to care about "invalid dependencies" vs "enforced dependencies"

- All rules could be made autofixable unless they conflict

- In a case of conflict, the interactive mode could be used to let the user figure out the right fix

**Describe the drawbacks of your solution**

I might be missing some cases where specifying invalid dependencies without having conflicts is actually a good behavior. I don't really see any, though 🤔

**Comment by arcanis** [comment]
Ping @bgotink, what do you think? Does my line of thought make sense?

**Comment by arcanis** [comment]
To give an example, this is the rule that prevents two workspaces from depending on different ranges:

```
gen_invalid_dependency(WorkspaceCwd, DependencyIdent, DependencyType, 'This dependency conflicts with another one from another workspace') :-
  workspace_has_dependency(WorkspaceCwd, DependencyIdent, DependencyRange, DependencyType),
  workspace_has_dependency(_, DependencyIdent, DependencyRange2, _),
  DependencyRange \= DependencyRange2.
```

With those changes, it would become as such:

```
gen_workspace_dependency_requirement(WorkspaceCwd, DependencyIdent, DependencyRange, DependencyType) :-
  workspace_has_dependency(WorkspaceCwd, DependencyIdent, _, DependencyType),
  workspace_has_dependency(_, DependencyIdent, DependencyRange, _).
```

**Comment by bgotink** [comment]
I haven't used `gen_invalid_dependency` yet, `gen_enforced_dependency_range` always fit my needs.

That said, the `constraints check` and `constraints fix` commands would have to become a lot smarter.

In the example you give, running `yarn constraints check` in a workspace with

```json5
// packages/foo/package.json
{
  "dependencies": {
    "a-dep": "1.0.0"
  }
}
// packages/bar/package.json
{
  "dependencies": {
    "a-dep": "2.0.0"
  }
}
```

would print

```
➤ YN0024: @scope/foo must depend on a-dep via npm:2.0.0 in dependencies, but uses npm:1.0.0 instead
➤ YN0024: @scope/bar must depend on a-dep via npm:1.0.0 in dependencies, but uses npm:2.0.0 instead
```

This is obviously confusing. It looks like we have to swap the two dependency versions, but that's definitely not what we want.

**Comment by arcanis** [comment]
Re-evaluation after each fix pass (or user input for conflicting ones) sounds like a good idea anyway. That's what `eslint --fix` actually does and it seems a good strategy when fixing one rule might lead to another error to appear. In this context it's more important to fix right than fix fast.

> In the example you give, running `yarn constraints check` in a workspace with [...] would print [...]

Yep, that would require a slight tweak to become something like this:

```
➤ YN0024: @scope/foo must depend on a-dep in dependencies via conflicting requirements of both npm:1.0.0* and npm:2.0.0
➤ YN0024: @scope/bar must depend on a-dep in dependencies via conflicting requirements of both npm:1.0.0 and npm:2.0.0*
```

**Comment by arcanis** [comment]
Fixed by #338
